ipywidgets/widgets/trait_types/__init__.py

```diff
@@ -0,0 +1,185 @@
+            year=pydt.year,
+            month=pydt.month - 1,  # Months are 0-based indices in JS
+            date=pydt.day,
+            hours=pydt.hour,       # Hours, Minutes, Seconds and Milliseconds are
+            minutes=pydt.minute,   # plural in JS
+            seconds=pydt.second,
+            milliseconds=pydt.microsecond / 1000
+        )
+
+def datetime_from_json(js, manager):
+    """Deserialize a Python datetime object from json"""
+    if js is None:
+        return None
+    else:
+        return dt.datetime(
+            js['year'],
+            js['month'] + 1,  # Months are 1-based in Python
+            js['date'],
+            js['hours'],
+            js['minutes'],
+            js['seconds'],
+            js['milliseconds'] * 1000
+        )
+
+datetime_serialization = {
+    'from_json': datetime_from_json,
+    'to_json': datetime_to_json
+}
+
+class InstanceDict(traitlets.Instance):
+    """An instance trait which coerces a dict to an instance.
+    
+    This lets the instance be specified as a dict, which is used to initialize the instance.
+    
+    Also, we default to a trivial instance, even if args and kwargs is not specified."""
+
+    def validate(self, obj, value):
+        if isinstance(value, dict):
+            return super(InstanceDict, self).validate(obj, self.klass(**value))
+        else:
+            return super(InstanceDict, self).validate(obj, value)
+
+    def make_dynamic_default(self):
+        return self.klass(*(self.default_args or ()),
+                          **(self.default_kwargs or {}))
+
+
+class EDict(eventful.Eventful, traitlets.Dict):
+
+    event_map = {
+        'setitem': ('__setitem__', 'setdefault'),
+        'delitem': ('__delitem__', 'pop'),
+        'update': 'update',
+        'clear': 'clear',
+    }
+    type_name = 'edict'
+
+    def _before_update(self, inst, call):
+        new = call.args[0]
+        new.update(call.kwargs)
+        call_list = []
+        return self.redirect(
+            None, 'setitem', inst,
+            args=new.items())
+
+    def _before_clear(self, inst, call):
+        return self.redirect(
+            None, 'delitem', inst,
+            args=inst.keys())
+
+    @staticmethod
+    def before_setitem(inst, call):
+        """Expect call.args[0] = key"""
+        key, old = call.args[0], inst.get(call.args[0], traitlets.Undefined)
+        def after_setitem(returned):
+            new = inst.get(key, traitlets.Undefined)
+            if not equivalent(old, new):
+                return traitlets.Bunch(key=key, old=old, new=new)
+        return after_setitem
+
+    _before_setitem = before_setitem
+    _before_delitem = before_setitem
+
+
+class EList(eventful.Eventful, traitlets.List):
+
+    event_map = {
+        'append': 'append',
+        'extend': 'extend',
+        'setitem': '__setitem__',
+        'reverse': 'reverse',
+        'sort': 'sort',
+    }
+    type_name = 'elist'
+
+    def before_setitem(self, inst, index):
```

Comment:
This should be static, and similarly to `EList` I think I'll just merge this into `_before_setitem`